src/smt/mam.cpp

```diff
@@ -1893,7 +1893,7 @@ namespace {
         void update_max_generation(enode * n, enode * prev) {
             m_max_generation = std::max(m_max_generation, n->get_generation());
 
-            if (m.has_trace_stream())
+            if (m.has_trace_stream() || is_trace_enabled("causality"))
```

Comment:
There are two kinds of logs: the log associated with debug tracing and the log associated with has_trace_stream().
In this PR you use the debug trace log for quantifier information. This then takes a dependency of compiling with TRACE enabled. The release builds don't have TRACE enabled. Is this sustainable for your use? 
There could be different modalities: (1) is some logging mechanism enabled, (2) where is it logged. 
These modalities are currently conflated, but your PR suggests to give it some thought whether conflating these will work for your use.

Reply:
Hi Nikolaj, thank you for your comments. In order to have these features available in the release build, the _is_trace_enabled("causality")_ should become something like _m.has_quantifier_trace_stream()_. 
At the same time, the tags "causality", "bindings", etc. should become input options to Z3 (e.g. causality=true or bindings=true).
I would like to introduce a new "quantifiers" trace that is independent from the existing trace. This new quantifiers trace is enabled as soon as one of the options (e.g. causality=true or bindings=true) is given in the command line.
In case this direction sounds good I am happy to work on it.

Reply:
It is up to you. At present it will only work if you build with the debug trace enabled. You can do this for release builds. It results in a slight overhead.